plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/procedure/MigrateProcedure.java

```diff
@@ -309,12 +315,14 @@ public Map<String, Optional<Partition>> listAllPartitions(HiveMetastore metastor
     }
 
     private List<DataFile> buildDataFiles(ConnectorSession session, RecursiveDirectory recursive, HiveStorageFormat format, String location, PartitionSpec partitionSpec, StructLike partition, NameMapping nameMapping)
-            throws IOException
+            throws IOException, ExecutionException, InterruptedException
     {
-        // TODO: Introduce parallelism
         TrinoFileSystem fileSystem = fileSystemFactory.create(session);
         FileIterator files = fileSystem.listFiles(location);
         ImmutableList.Builder<DataFile> dataFilesBuilder = ImmutableList.builder();
+        int numThreads = Runtime.getRuntime().availableProcessors();
+        ExecutorService executor = Executors.newFixedThreadPool(numThreads);
```

Comment:
The thread pool should be Guice-managed.
See example in Hive
https://github.com/trinodb/trino/blob/b6640e96aa4e61f5eb06e9660ace5893aade3b95/plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveModule.java#L167-L172
https://github.com/trinodb/trino/blob/2a88bd74e37579484d259f37389c4cb1da5484c8/plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveSplitManager.java#L139